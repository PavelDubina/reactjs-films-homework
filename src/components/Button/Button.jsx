import React, { Children } from 'react'
import PropTypes from 'prop-types'

import './Button.scss'

const Button = ({ styles, onViewInfo, children }) => {
  return (
    <button className={styles} onClick={onViewInfo}>
      {children}
    </button>
  )
}

export default Button
