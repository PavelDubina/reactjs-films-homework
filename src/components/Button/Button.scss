button {
  width: 110px;
  height: 35px;
  padding: 5px;
  margin: 0 10px;
  border: none;
  border-radius: 5px;
  outline: none;
  font-size: 16px;
  cursor: pointer;

  &:active {
    opacity: 0.95;
  }
}
