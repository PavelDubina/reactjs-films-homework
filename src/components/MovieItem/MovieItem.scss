/*
 * Copyright Â© 2020 EPAM Systems, Inc. All Rights Reserved. All information contained herein is, and remains the
 * property of EPAM Systems, Inc. and/or its suppliers and is protected by international intellectual
 * property law. Dissemination of this information or reproduction of this material is strictly forbidden,
 * unless prior written permission is obtained from EPAM Systems, Inc
 */

.container {
  position: relative;
  display: flex;
  flex-direction: column;
  width: 100%;
  height: 100%;
  border-radius: 10px;
  background-color: #fff;
}

.poster {
  flex: 3;
  position: relative;
  border-top-left-radius: 10px;
  border-top-right-radius: 10px;
  background-size: 100%;
  background-color: grey;

  &__hover {
    display: none;
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    background-color: rgba(0, 0, 0, 0.5);
    border-top-left-radius: 10px;
    border-top-right-radius: 10px;

    &_play {
      font-size: 84px;
      color: rgb(24, 183, 236);
      background-color: #fff;
      border-radius: 50%;
      cursor: pointer;
    }

    &_text {
      margin: 10px 0;
      font-size: 20px;
      line-height: 28px;
      color: #fff;
    }

    &_button {
      margin: 10px 0;
      background-color: transparent;
      border: 1px solid #fff;
      color: #fff;
    }
  }

  &:hover &__hover {
    display: flex;
  }
}
