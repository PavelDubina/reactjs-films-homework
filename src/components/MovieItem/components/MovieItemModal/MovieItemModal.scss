/*
 * Copyright Â© 2020 EPAM Systems, Inc. All Rights Reserved. All information contained herein is, and remains the
 * property of EPAM Systems, Inc. and/or its suppliers and is protected by international intellectual
 * property law. Dissemination of this information or reproduction of this material is strictly forbidden,
 * unless prior written permission is obtained from EPAM Systems, Inc
 */

.modal {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  padding: 10px;
  background-color: rgba(0, 0, 0, 0.8);
  border-radius: 10px;
  z-index: 50;

  &__window {
    position: relative;
    padding: 10px;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
  }

  &__cross {
    position: absolute;
    top: 0;
    left: 0;
    color: #fff;
    font-size: 24px;
    cursor: pointer;
  }

  &__description {
    margin: 10px 0;
    color: #fff;
    height: 200px;
    overflow-y: auto;
  }

  &__button {
    align-self: flex-start;
    margin: 0;
    background-color: rgb(24, 183, 236);
    color: #fff;
  }
}

.info {
  flex: 1;
  width: 100%;

  &__about {
    padding: 30px 0 5px 0;
    display: flex;
    justify-content: space-between;

    &_title {
      color: #fff;
      font-size: 24px;
      padding-right: 10px;
      font-weight: bold;
      overflow: hidden;
      white-space: nowrap;
      text-overflow: ellipsis;
    }

    &_vote {
      font-weight: bold;
      color: rgb(24, 183, 236);
      padding: 1px 7px;
      border: 2px solid rgb(24, 183, 236);
      border-radius: 5px;
    }
  }

  &__genres {
    color: rgb(24, 183, 236);
    font-weight: 800;
  }
}
