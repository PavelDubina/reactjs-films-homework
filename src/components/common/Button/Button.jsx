import React from 'react'
import PropTypes from 'prop-types'
import className from 'classnames'
import styles from './Button.scss'

export const Button = ({ type = 'primary', onClick, children }) => {
  const classList = className(styles.button, {
    [styles.primary]: type === 'primary',
    [styles.secondary]: type === 'secondary',
    [styles.info]: type === 'info',
  })
  return (
    <button className={classList} onClick={onClick}>
      {children}
    </button>
  )
}

Button.propTypes = {
  type: PropTypes.string,
  onClick: PropTypes.func.isRequired,
  children: PropTypes.node.isRequired,
}
