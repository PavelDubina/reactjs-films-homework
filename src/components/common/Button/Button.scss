.button {
  width: 110px;
  height: 35px;
  padding: 5px;
  margin: 0 10px;
  border: none;
  border-radius: 5px;
  outline: none;
  font-size: 16px;
  cursor: pointer;

  &:active {
    opacity: 0.95;
  }

  @media (max-width: 599px) {
    height: 32px;
    width: 100px;
    margin: 0 5px;
  }
}

.primary {
  background-color: rgb(24, 183, 236);
  color: #fff;
}

.secondary {
  background-color: inherit;
  border: 1px solid #fff;
  color: #fff;
}

.info {
  background-color: #f0f0f0;
  color: rgb(24, 183, 236);
}
