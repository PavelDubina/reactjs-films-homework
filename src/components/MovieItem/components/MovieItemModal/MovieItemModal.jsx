/*
 * Copyright Â© 2020 EPAM Systems, Inc. All Rights Reserved. All information contained herein is, and remains the
 * property of EPAM Systems, Inc. and/or its suppliers and is protected by international intellectual
 * property law. Dissemination of this information or reproduction of this material is strictly forbidden,
 * unless prior written permission is obtained from EPAM Systems, Inc
 */

import React from 'react'
import PropTypes from 'prop-types'
import { faTimesCircle } from '@fortawesome/free-regular-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { useDispatch } from 'react-redux'
import { fetchTrailerData } from '../../../../redux/actions/fetchTrailer'
import { Info } from '../../../common/Info/Info'
import { Button } from '../../../common/Button/Button'
import styles from './MovieItemModal.scss'

export const MovieItemModal = ({ genreIds, id, overview, title, vote, onModal }) => {
  const dispatch = useDispatch()
  return (
    <div className={styles.modal}>
      <div className={styles.modal__window}>
        <FontAwesomeIcon onClick={onModal} className={styles.modal__cross} icon={faTimesCircle} />
        <Info type="info-modal" genreIds={genreIds} title={title} vote={vote} />
        <div className={styles.modal__description}>{overview}</div>
        <div className={styles.modal__button}>
          <Button
            onClick={() => {
              dispatch(fetchTrailerData(id))
              onModal()
            }}
            type="primary"
          >
            Watch Now
          </Button>
        </div>
      </div>
    </div>
  )
}

MovieItemModal.propTypes = {
  onModal: PropTypes.func.isRequired,
  vote: PropTypes.number.isRequired,
  genreIds: PropTypes.arrayOf(PropTypes.number).isRequired,
  overview: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  id: PropTypes.number.isRequired,
}
