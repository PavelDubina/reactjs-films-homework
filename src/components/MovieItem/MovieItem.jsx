/*
 * Copyright Â© 2020 EPAM Systems, Inc. All Rights Reserved. All information contained herein is, and remains the
 * property of EPAM Systems, Inc. and/or its suppliers and is protected by international intellectual
 * property law. Dissemination of this information or reproduction of this material is strictly forbidden,
 * unless prior written permission is obtained from EPAM Systems, Inc
 */

import React, { useState } from 'react'
import { useDispatch } from 'react-redux'
import PropTypes from 'prop-types'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faPlayCircle } from '@fortawesome/free-solid-svg-icons'
import Button from '../common/Button'
import Info from '../common/Info'
import MovieItemModal from './components/MovieItemModal'
import { fetchFilmData } from '../../redux/actions/fetchFilmData'
import { fetchTrailerData } from '../../redux/actions/fetchTrailer'
import styles from './MovieItem.scss'

const MovieItem = ({ film }) => {
  const dispatch = useDispatch()
  const [isModal, setIsModal] = useState(false)
  const onModal = (setState) => setState((state) => !state)
  return (
    <div className={styles.container}>
      {isModal && <MovieItemModal onModal={() => onModal(setIsModal)} film={film} />}
      <div
        className={styles.poster}
        style={{ backgroundImage: `url(https://image.tmdb.org/t/p/original${film.poster_path})` }}
      >
        <div className={styles.poster__hover}>
          <FontAwesomeIcon
            onClick={() => dispatch(fetchTrailerData(film.id))}
            className={styles.poster__hover_play}
            icon={faPlayCircle}
          />
          <span className={styles.poster__hover_text}>Watch Now</span>
          <Button onClick={() => onModal(setIsModal)} className={styles.poster__hover_button}>
            View Info
          </Button>
        </div>
      </div>
      <Info onClick={() => dispatch(fetchFilmData(film.id))} className={styles} film={film} />
    </div>
  )
}

MovieItem.propTypes = {
  film: PropTypes.object.isRequired,
}

export default MovieItem
