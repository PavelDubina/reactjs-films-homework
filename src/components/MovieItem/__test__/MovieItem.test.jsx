/*
 * Copyright Â© 2020 EPAM Systems, Inc. All Rights Reserved. All information contained herein is, and remains the
 * property of EPAM Systems, Inc. and/or its suppliers and is protected by international intellectual
 * property law. Dissemination of this information or reproduction of this material is strictly forbidden,
 * unless prior written permission is obtained from EPAM Systems, Inc
 */

import React from 'react'
import { create, act } from 'react-test-renderer'
import { MemoryRouter } from 'react-router-dom'
import configureStore from 'redux-mock-store'
import { Provider } from 'react-redux'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { MovieItem } from '../MovieItem'
import { Info } from '../../common/Info/Info'
import { Button } from '../../common/Button/Button'
import { MovieItemModal } from '../components/MovieItemModal/MovieItemModal'

const mockHistoryPush = jest.fn()

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}))

describe('MovieItem', () => {
  const mockStore = configureStore([])
  let store
  beforeEach(() => {
    store = mockStore({
      genres: { genres: [1, 2, 3] },
    })
    store.dispatch = jest.fn()
  })

  it('Should render and match snapshot, go to route details, dispatch trailer data', () => {
    const setIsModal = jest.fn()
    const handleClick = jest.spyOn(React, 'useState')
    handleClick.mockImplementation((isModal) => [isModal, setIsModal])
    const component = create(
      <MemoryRouter>
        <Provider store={store}>
          <MovieItem overview="test" vote={5} title="title" genreIds={[28, 12, 16]} poster="poster" id={1} />
        </Provider>
      </MemoryRouter>,
    )
    act(() => {
      component.root.findByType(Info).props.onClick()
    })
    expect(component.toJSON()).toMatchSnapshot()
    expect(mockHistoryPush).toHaveBeenCalledWith('/details/1')

    act(() => {
      component.root.findByType(FontAwesomeIcon).props.onClick()
    })
    expect(store.dispatch).toHaveBeenCalledTimes(1)
    act(() => {
      component.root.findByType(Button).props.onClick()
    })
    expect(setIsModal).toBeTruthy()
    act(() => {
      component.root.findByType(MovieItemModal).props.onModal()
    })
    expect(setIsModal).toBeTruthy()
  })
})
