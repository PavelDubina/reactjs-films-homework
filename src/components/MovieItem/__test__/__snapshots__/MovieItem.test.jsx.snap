// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieItem Should render and match snapshot, go to route details, dispatch trailer data 1`] = `
<div
  className="container"
>
  <div
    className="poster"
    style={
      Object {
        "backgroundImage": "url(https://image.tmdb.org/t/p/originalposter)",
      }
    }
  >
    <div
      className="poster__hover"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-play-circle fa-w-16 poster__hover_play"
        data-icon="play-circle"
        data-prefix="fas"
        focusable="false"
        onClick={[Function]}
        role="img"
        style={Object {}}
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm115.7 272l-176 101c-15.8 8.8-35.7-2.5-35.7-21V152c0-18.4 19.8-29.8 35.7-21l176 107c16.4 9.2 16.4 32.9 0 42z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
      <span
        className="poster__hover_text"
      >
        Watch Now
      </span>
      <button
        className="button secondary"
        onClick={[Function]}
      >
        View Info
      </button>
    </div>
  </div>
  <div
    className="info"
    onClick={[Function]}
  >
    <div
      className="info__about"
    >
      <div
        className="info__about_title"
      >
        title
      </div>
      <div
        className="vote"
      >
        5
      </div>
    </div>
    <div
      className="info__genres"
    >
      
    </div>
  </div>
</div>
`;
